samtranslator/validator/validator.py

```diff
@@ -1,55 +1,259 @@
-            # Swallowing expected exception here as our caller is expecting validation errors and
-            # not the valiation exception itself
-            pass
+        # Tree of Error objects
+        # Each object can have a list of child errors in its Context attribute
+        validation_errors = self.validator.iter_errors(template_dict)
 
-        return validation_errors
+        # Set of "[Path.To.Element] Error message"
+        # To track error uniqueness, Dict instead of List, for speed
+        errors_set = {}
 
-    @staticmethod
-    def _read_schema():
+        for e in validation_errors:
+            self._process_error(e, errors_set)
+
+        # To be consistent across python versions 2 and 3, we have to sort the final result
+        # It seems that the validator is not receiving the properties in the same order between python 2 and 3
+        # It thus returns errors in a different order
+        return sorted(errors_set.keys())
+
+    def _process_error(self, error, errors_set):
         """
-        Reads the JSON Schema at given file path
+        Processes the validation errors recursively
+        error is actually a tree of errors
+        Each error can have a list of child errors in its 'context' attribute
 
-        :param string schema_file: Optional path to the schema file. If not provided, the system configured value
-            will be used
-        :return dict: JSON Schema of the policy template
+        Parameters
+        ----------
+        error : Error
+            Error at the head
+        errors_set : Dict
+            Set of formatted errors
         """
-        return SamTemplateValidator._read_json(sam_schema.SCHEMA_FILE)
+        if error is None:
+            return
 
-    @staticmethod
-    def _read_json(filepath):
+        if not error.context:
+            # We only display the leaves
+            # Format the message with pseudo JSON Path:
+            # [Path.To.Element] Error message
+            error_path = ".".join([str(p) for p in error.absolute_path]) if error.absolute_path else "."
+
+            error_content = "[{}] {}".format(error_path, self._cleanup_error_message(error))
+
+            if error_content not in errors_set:
+                # We set the value to None as we don't use it
+                errors_set[error_content] = None
+            return
+
+        for context_error in error.context:
+            # Each "context" item is also a validation error
+            self._process_error(context_error, errors_set)
+
+    def _cleanup_error_message(self, error):
+        """
+        Cleans an error message up to remove unecessary clutter or replace
+        it with a more meaningful one
+
+        Parameters
+        ----------
+        error : Error
+            Error message to clean
+
+        Returns
+        -------
+        str
+            Cleaned message
+        """
+        final_message = re.sub(self.UNICODE_TYPE_REGEX, r"\1", error.message)
```

Comment:
A bit confused on this part. So `error.message` is a string that contains a unicode prefixed string?
`error.message == " Error message: u'inner string' "` 
Will this break anything if a non-ascii character is contained inside of the inner string in Python 2?

Reply:
When run on Python2, the `jsonschema` validator returns error messages with `u` prefixed strings.
For example: `"[Resources] 3 is not of type u'object'"`.
On Python3 it returns: `"[Resources] 3 is not of type 'object'"`.
We need to remove the `u` to have tests results consistent between py2 and py3.
(`u`'s are also ugly)